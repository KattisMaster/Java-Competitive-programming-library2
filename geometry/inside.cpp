/* KTH ACM Contest Template Library
 *
 * Geometry/Polygons/Inside
 *
 * Credit:
 *   By Per Austrin
 */

#include "point_line_relations.cc"

template <class It, class P>
bool poly_inside(It begin, It end, const P &p, bool strict = true) {
  bool inside = false;
  for (It i = begin, j = end - 1; i != end; j = i++) {
    if (on_segment(*j, *i, p)) return !strict;
    if (min(j->x, i->x) < p.x && max(j->x, i->x) >= p.x &&
	abs(j->x - i->x)*(p.y - i->y) > abs(p.x - i->x)*(j->y - i->y))
      inside ^= 1;
  }
  return inside;
}
