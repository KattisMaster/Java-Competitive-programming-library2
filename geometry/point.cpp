/* KTH ACM Contest Template Library
 *
 * Geometry/Primitives/Point structure
 *
 * Credit:
 *   By Mattias de Zalenski
 */
#include <cmath>

template <class T>
struct point {
  typedef T coord_type;
  T x, y; point(T _x=T(), T _y=T()) : x(_x), y(_y) { }
  bool operator<(const point<T> &p) const { return x!=p.x ? x<p.x : y<p.y; }
  point<T> operator-(const point<T> &p)const { return point<T>(x-p.x,y-p.y); }
  point<T> operator/(T d) const { return point<T>(x/d, y/d); }
};


template <class T> T dX(T p, T q) { return q - p; }

template <class T> T dx(point<T> p, point<T> q) { return q.x - p.x; }
template <class T> T dy(point<T> p, point<T> q) { return q.y - p.y; }

template <class T> T sqr(T x) { return x * x; }
template <class T> T dist2(point<T> p) { return sqr(p.x) + sqr(p.y); }
template <class T> T dist(point<T> p) { return sqrt((double) dist2(p)); }

template <class T> T dot(point<T> p, point<T> q) { return p.x*q.x + p.y*q.y; }
template <class T> T cross(point<T> p, point<T> q) { return p.x*q.y-p.y*q.x; }

template <class P> P unit(P p) { return p / dist(p); }
template <class P> P perp(P p) { return P(-p.y, p.x); }
template <class P> P normal(P p) { return unit(perp(p)); }
