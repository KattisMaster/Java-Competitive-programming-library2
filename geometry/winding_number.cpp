/*****************************************************************************
 * 
 * Geometry1: inside_poly
 * ======================
 *
 * Determine if a point is inside/outside or on the edge of a polygon.
 *
 * INPUT
 * -----
 * p:    An array of size n representing the polygon, given as points either
 *       in cw or ccw order. (the start/end point should not be repeated)
 * n:    the number of vertices
 * t:    The test point
 *
 * OUTPUT
 * ------
 * +1    inside
 * 0     on the edge
 * -1    outside
 *
 * COMPLEXITY  O( n )
 *
 * REQUIRES  geometry.h, pointline.cpp (onedge)
 * ------------------------------------------------------------------------- *
 *
 * NADA acmlib (10 March 2002)
 * Templates for KTH-NADA, Ballons'R'Us, ACM 2001-2002
 *   Swedish competition, Link?ping 6 Oct 2001
 *   Revised for SWERC, Portu, Portugal 17 Nov 2001
 *   Revised for World Finals, Honolulu, 23 Mar 2002
 *   David Rydh, Mattias de Zalenski, Fredrik Niemel?
 *
 *****************************************************************************/

#include "pointline.cpp"

template<class V, class P>
double winding_nr(const V &p, int n, P t, bool &onEdge) {
  double wind = 0;

  onEdge = false;

  for (int i=0, j=n-1; i<n; j=i++) {
    if( onsegment(p[i], p[j], t) ) {
      onEdge = true;
      continue;
    }
    double t1 = theta(t-p[i]), t2 = theta(t-p[j]);
    double dt = t1-t2;
    if( dt > 2 ) dt -= 4;
    if( dt < -2 ) dt += 4;
    wind += dt;
  }

  return wind;
}

template<class P, class V>     // V is vector/array of point<T>s
int inside_poly(P t, const V &p, int n) {
  bool edge;
  double wind = winding_nr(p,n,t, edge);

  if( edge ) return wind>4 ? 1:0;

  // Not on edge, i.e. wind is (approx) 4*nr of turns.
  return wind > 2 ? 1:-1;
}
