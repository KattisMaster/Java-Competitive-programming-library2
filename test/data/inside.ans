 origo inside unit square
inside
inside
 points outside unit square
outside
outside

outside
outside

outside
outside

outside
outside

outside
outside

outside
outside
 points on unit square N border
outside
on
 points on unit square E border
outside
on
 points on unit square S border
inside
on
 points on unit square W border
inside
on
 points on unit square NE border
outside
on
 points on unit square SE border
outside
on
 points on unit square SW border
inside
on
 points on unit square NW border
outside
on
 points outside rectangular polygon but inside convex hull
outside
outside

outside
outside
