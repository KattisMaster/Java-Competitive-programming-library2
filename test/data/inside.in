% origo inside unit square
0 0
4
1 -1
1 1
-1 1
-1 -1
% points outside unit square
2 3
4
1 -1
1 1
-1 1
-1 -1
%
4 8
4
1 -1
1 1
-1 1
-1 -1
%
6 4
4
1 -1
1 1
-1 1
-1 -1
%
2 9
4
1 -1
1 1
-1 1
-1 -1
%
0 7
4
1 -1
1 1
-1 1
-1 -1
%
6 0
4
1 -1
1 1
-1 1
-1 -1
% points on unit square N border
0 1
4
1 -1
1 1
-1 1
-1 -1
% points on unit square E border
1 0
4
1 -1
1 1
-1 1
-1 -1
% points on unit square S border
0 -1
4
1 -1
1 1
-1 1
-1 -1
% points on unit square W border
-1 0
4
1 -1
1 1
-1 1
-1 -1
% points on unit square NE border
1 1
4
1 -1
1 1
-1 1
-1 -1
% points on unit square SE border
1 -1
4
1 -1
1 1
-1 1
-1 -1
% points on unit square SW border
-1 -1
4
1 -1
1 1
-1 1
-1 -1
% points on unit square NW border
-1 1
4
1 -1
1 1
-1 1
-1 -1
% points outside rectangular polygon but inside convex hull
3 3
8
1 1
1 5
4 5
4 4
2 4
2 2
4 2
4 1
%
5 3
12
1 1
1 6
7 6
7 1
3 1
3 4
4 4
4 2
6 2
6 5
2 5
2 1
