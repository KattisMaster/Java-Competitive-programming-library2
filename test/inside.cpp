#include "Testplate.cpp"

using namespace std;

#include "../geometry/point.cpp"
//#include "../geometry/point_ops.cpp"
#include "../geometry/inside.cpp"
#include "../geometry/winding_number.cpp"

typedef point<double> Pt;

void init() {
}

// In:
// px py
// n
// qx qy
// ...(n)

// Out:
// inside / outside (inside.cpp)
// inside / on / outside (winding_number.cpp)

bool solve(int P) {
  Pt q;
  cin >> q.x >> q.y;
  int n;
  cin >> n;
  if (cin.fail() || n == 0) return false;
  vector<Pt> v;
  for (int i = 0; i < n; ++i) {
    Pt p; cin >> p.x >> p.y;
    v.push_back(p);
  }
  if (inside(v, n, q))
    cout << "inside" << endl;
  else
    cout << "outside" << endl;

  int wind = inside_wn(v, n, q);
  switch (wind) {
  case 1: cout << "inside" << endl; break;
  case 0: cout << "on" << endl; break;
  case -1: cout << "outside" << endl; break;
  }
  return true;
}
