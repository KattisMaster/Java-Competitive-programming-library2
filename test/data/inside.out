 origo inside unit square
inside
 points outside unit square
outside

outside

outside

outside

outside

outside
 points on unit square N border
on
 points on unit square E border
on
 points on unit square S border
on
 points on unit square W border
on
 points on unit square NE border
on
 points on unit square SE border
on
 points on unit square SW border
on
 points on unit square NW border
on
 points outside polygon but inside convex hull
outside

outside